src/main/scala/org/constellation/storage/SnapshotService.scala

```diff
@@ -273,36 +272,50 @@ class SnapshotService[F[_]: Concurrent](
+  private[storage] def applySnapshot()(implicit C: ContextShift[F]): EitherT[F, SnapshotError, Unit] = {
+    val write: Snapshot => F[Either[SnapshotError, Unit]] = (currentSnapshot: Snapshot) =>
+      writeSnapshotToDisk(currentSnapshot)
+        .flatTap(
+          _ =>
+            applyAfterSnapshot(currentSnapshot)
+              .flatTap(_ => dao.metrics.incrementMetricAsync(Metrics.snapshotCount))
+        )
+
+    EitherT {
+      snapshot.get.flatMap { currentSnapshot =>
+        if (currentSnapshot == Snapshot.snapshotZero) Sync[F].pure[Either[SnapshotError, Unit]](Right(()))
+        else write(currentSnapshot)
       }
     }
+  }
 
-  def addSnapshotToDisk(snapshot: StoredSnapshot): F[Path] =
+  def addSnapshotToDisk(snapshot: StoredSnapshot): F[Either[Throwable, Path]] =
     Snapshot.writeSnapshot[F](snapshot)
 
-  private def writeSnapshotToDisk(currentSnapshot: Snapshot)(implicit C: ContextShift[F]): F[Unit] =
+  private def writeSnapshotToDisk(
+    currentSnapshot: Snapshot
+  )(implicit C: ContextShift[F]): F[Either[SnapshotError, Unit]] =
     currentSnapshot.checkpointBlocks.toList
-      .traverse(checkpointService.fullData)
+      .traverse(h => checkpointService.fullData(h).map(d => (h, d)))
       .flatMap {
         case maybeBlocks
-            if maybeBlocks.exists(maybeCache => maybeCache.isEmpty || maybeCache.exists(_.checkpointBlock.isEmpty)) =>
-          // TODO : This should never happen, if it does we need to reset the node state and redownload
-          dao.metrics
-            .incrementMetricAsync("snapshotWriteToDiskMissingData")
-            .flatMap(_ => Sync[F].raiseError[Unit](new IllegalStateException("Snapshot data is missing")))
+            if maybeBlocks.exists(
+              maybeCache => maybeCache._2.isEmpty || maybeCache._2.exists(_.checkpointBlock.isEmpty)
+            ) =>
+          Sync[F]
+            .pure[Either[SnapshotError, Unit]] {
+              val maybeEmpty =
+                maybeBlocks.find(maybeCache => maybeCache._2.isEmpty || maybeCache._2.exists(_.checkpointBlock.isEmpty))
+              logger.error(s"Snapshot data is missing for block: ${maybeEmpty}")
+              Left(SnapshotIllegalState)
+            }
+            .flatTap(_ => dao.metrics.incrementMetricAsync("snapshotInvalidData"))
         case maybeBlocks =>
-          val flatten = maybeBlocks.flatten.sortBy(_.checkpointBlock.map(_.baseHash))
-          withMetric[F, Unit](
-            Snapshot.writeSnapshot(StoredSnapshot(currentSnapshot, flatten)).void,
-            Metrics.snapshotWriteToDisk
-          )
+          val flatten = maybeBlocks.flatMap(_._2).sortBy(_.checkpointBlock.map(_.baseHash))
+          Snapshot
+            .writeSnapshot(StoredSnapshot(currentSnapshot, flatten))
+            .flatTap(e => dao.metrics.incrementMetricAsync(Metrics.snapshotWriteToDisk, e))
+            .map(e => e.leftMap(ex => SnapshotIOError(ex)).map(p => logger.debug(s"Snapshot written at $p")))
```

Comment:
`.map(_.leftMap(SnapshotIOError(_)).map(p => ...`

Reply:
Can't find type after

Reply:
Weird. It works for me on your branch: 
<img width="693" alt="image" src="https://user-images.githubusercontent.com/4959931/67374029-eb12af00-f580-11e9-872c-4f8d0bd3bbd3.png">